crates/transpiler/src/passes/litinski_transformation.rs

```diff
@@ -0,0 +1,213 @@
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::imports::PAULI_EVOLUTION_GATE;
+use qiskit_circuit::operations::{
+    multiply_param, Operation, OperationRef, Param, PyGate, StandardGate,
+};
+use qiskit_circuit::packed_instruction::PackedInstruction;
+use qiskit_circuit::{Clbit, Qubit, VarsMode};
+
+use qiskit_quantum_info::clifford::Clifford;
+use qiskit_quantum_info::sparse_observable::PySparseObservable;
+
+use smallvec::smallvec;
+use std::f64::consts::PI;
+
+// List of gate names supported by the pass: the pass is skipped if the circuit
+// contains gate names outside of this list.
+const SUPPORTED_GATE_NAMES: &[&str; 19] = &[
+    "id", "x", "y", "z", "h", "s", "sdg", "sx", "sxdg", "cx", "cz", "cy", "swap", "iswap", "ecr",
+    "dcx", "t", "tdg", "rz",
+];
+
+// List of rotation gate names: the pass is skipped if the circuit contains
+// no gate names in this list.
+const ROTATION_GATE_NAMES: &[&str; 3] = &["t", "tdg", "rz"];
+
+/// Expresses a given circuit as a sequence of Pauli rotations followed by a final Clifford operator.
+pub fn extract_rotations(circuit: &[(String, Vec<usize>)], nqubits: usize) -> Vec<(bool, String)> {
+    let mut clifford = Clifford::identity(nqubits);
+    let mut rotations: Vec<(bool, String)> = Vec::new();
+
+    for (gate_name, qbits) in circuit.iter() {
+        match gate_name.as_str() {
+            "id" => {}
+            "x" => clifford.append_x(qbits[0]),
+            "y" => clifford.append_y(qbits[0]),
+            "z" => clifford.append_z(qbits[0]),
+            "h" => clifford.append_h(qbits[0]),
+            "s" => clifford.append_s(qbits[0]),
+            "sdg" => clifford.append_sdg(qbits[0]),
+            "sx" => clifford.append_sx(qbits[0]),
+            "sxdg" => clifford.append_sxdg(qbits[0]),
+            "cx" => clifford.append_cx(qbits[0], qbits[1]),
+            "cz" => clifford.append_cz(qbits[0], qbits[1]),
+            "cy" => clifford.append_cy(qbits[0], qbits[1]),
+            "swap" => clifford.append_swap(qbits[0], qbits[1]),
+            "iswap" => clifford.append_iswap(qbits[0], qbits[1]),
+            "ecr" => clifford.append_ecr(qbits[0], qbits[1]),
+            "dcx" => clifford.append_dcx(qbits[0], qbits[1]),
+            "rz" => {
+                rotations.push(clifford.get_inverse_z(qbits[0]));
+            }
+            _ => panic!("Unsupported gate {}", gate_name),
+        }
+    }
+
+    rotations
+}
+
+#[pyfunction]
+#[pyo3(name = "run")]
+pub fn run_litinski_transformation(
+    py: Python,
+    dag: &mut DAGCircuit,
+) -> PyResult<Option<DAGCircuit>> {
+    let op_counts = dag.get_op_counts();
+
+    // Skip the pass if there are no rotation gates.
+    if op_counts
+        .keys()
+        .all(|k| !ROTATION_GATE_NAMES.contains(&k.as_str()))
+    {
+        return Ok(None);
+    }
+
+    // Skip the pass if there are unsupported gates.
+    if !op_counts
+        .keys()
+        .all(|k| SUPPORTED_GATE_NAMES.contains(&k.as_str()))
```

Comment:
Tiny nit; but I think calling the `any` version should be faster in theory as the iterator can stop once the clause is satisfied once:
```suggestion
    if op_counts
        .keys()
        .any(|k| !SUPPORTED_GATE_NAMES.contains(&k.as_str()))
```

Reply:
Nitpick rejected 😄. Both `all` and `any` short-circuit: `any` stops when it sees `True`, and `all` stops when it sees `False`.

Here is some simple Rust code that you can run to verify this:
```
fn main() {
    let v = vec![1, 3, 5, 6, 9, 11];
    
    let all_are_odd = v.iter().all(|z| {
        println!("Examining {:?}", z);
        z % 2 != 0}
    );

    println!("all_are_odd = {}", all_are_odd);
}
```

Reply:
True that makes sense -- I don't know what I was thinking 😂